derive/src/lib.rs

```diff
@@ -140,21 +141,75 @@ pub fn decode_derive(input: TokenStream) -> TokenStream {
 		}
 	};
 
-	let mut new_name = "_IMPL_DECODE_FOR_".to_string();
-	new_name.push_str(name.to_string().trim_start_matches("r#"));
-	let dummy_const = Ident::new(&new_name, Span::call_site());
-	let parity_codec_crate = include_parity_scale_codec_crate();
+	wrap_with_dummy_const(&input, "_IMPL_DECODE_FOR_", impl_block)
+}
 
-	let generated = quote! {
-		#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
-		const #dummy_const: () = {
-			#[allow(unknown_lints)]
-			#[cfg_attr(feature = "cargo-clippy", allow(useless_attribute))]
-			#[allow(rust_2018_idioms)]
-			#parity_codec_crate
-			#impl_block
-		};
+#[proc_macro_derive(CompactAs, attributes(codec))]
+pub fn decode_compact_as(input: TokenStream) -> TokenStream {
+	let mut input: DeriveInput = match syn::parse(input) {
+		Ok(input) => input,
+		Err(e) => return e.to_compile_error().into(),
 	};
 
-	generated.into()
+	if let Err(e) = trait_bounds::add(
+		&input.ident,
+		&mut input.generics,
+		&input.data,
+		parse_quote!(_parity_scale_codec::CompactAs),
+		None,
+	) {
+		return e.to_compile_error().into();
+	}
+
+	let name = &input.ident;
+	let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+	let (inner_ty, inner_field, constructor) = match input.data {
+		Data::Struct(ref data) => {
+			match data.fields {
+				Fields::Named(ref fields) if fields.named.iter().count() == 1 => {
+					let field = fields.named.iter().next().expect("Exactly one field");
+					let field_name = &field.ident;
+					(&field.ty, quote!(&self.#field_name), quote!(#name { #field_name: x }))
+				},
+				Fields::Unnamed(ref fields) if fields.unnamed.iter().count() == 1 => {
+					let (id, field) = fields.unnamed.iter().enumerate().next().expect("Exactly one field");
+					let id = syn::Index::from(id);
+					(&field.ty, quote!(&self.#id), quote!(#name(x)))
+				},
+				_ => {
+					return Error::new(
+						Span::call_site(),
+						"Only structs with a single field can derive CompactAs"
```

Comment:
Maybe we should also take into consideration the skip attribute here, and instead only struct with only a single not skipped field can derive CompactAs

Reply:
Yeah, @thiolliere has already written some code for the `using_encoded` optimizations.

Reply:
https://github.com/paritytech/parity-scale-codec/pull/97/commits/9e24127cda7d1756dd90e0758386d726d639231e